### PR TITLE
feat: add 'mousescroll' option

### LINKED ISSUE #11738: Ability to set number of lines to scroll on mouse wheel input
- `nvim --version`: NVIM v0.5.0-317-gfd89ad7bf
- `vim -u DEFAULTS` (version: ) behaves differently? no
- Operating system/version: linux, kernel 5.4.12
- Terminal name/version: alacritty 0.4.1
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`
```
nvim -u NORC
:set mouse=a
:e ~/some_long_file
<move mousewheel one tick>
```

### Actual behaviour
Scroll amount is hardcoded/fixed at 3 lines

### Expected behaviour
Scroll amount should be configurable

### Use case
Touchpad scrolling, having the scroll quantum set to 1 allows for visually much smoother scrolling

### See-also
https://github.com/szbergeron/neovim, all visible sites where value is hardcoded changed to different value (should be made configurable for any final patch)

### PR DESCRIPTION
**Pull request by JaySandhu**
Building on https://github.com/neovim/neovim/pull/1288 by @cepko33  and on https://github.com/neovim/neovim/pull/10954 by @qstrahl.
Close #11738

Adds a mousescroll option which I understand now has the green light.

**Motivation**
Scrolling with a trackpad feels off with the current hardcoded value. Remapping the scroll wheel keys to \<C-Y> and \<C-E>, as is commonly suggested, is not a great alternative, you can no longer scroll inactive windows.

The new option defaults to 1, which is different from the old hardcoded value of 3, but I think is a more reasonable default. This can be changed if needed.

Horizontal scrolling still uses a hardcoded value (6 I believe).

- [x] Add the option
- [x] Document the option
- [x] Test the option manually
- [x] Create a new automated test
- [x] Update `:help scroll-mouse-wheel`
- [x] Update `:help vim-differences`

### CODE REVIEW

**Comment by justinmk** [comment]
Thanks for picking this up. Needs some tests; I added the checklist from https://github.com/neovim/neovim/pull/10954 to your description.

> Horizontal scrolling still uses a hardcoded value (6 I believe).

Will we need something like...

    :set mousescroll=vert:42,hor:42

**Comment by JaySandhu** [comment]
> Will we need something like...
>
> ```
> :set mousescroll=ver:42,hor:42
> ```

I like that idea. It would make for a more complete solution. I'll rework the PR tomorrow.

Do we want to force users to set both directions? Or should we allow them to omit either `ver` or `hor` (in which case we can fallback to some default value). To put my two cents in, I think that we should support omitted directions.

**Comment by JaySandhu** [comment]
I gave it a shot. The new approach is more complete, but also, more complex. The `mousescroll` option is now a string in the following format:

```
set mousescroll=direction:count,direction:count
```

Where direction is either 'ver' or 'hor', and count is a non negative integer.

Some points that may be contentious:

* Setting a direction twice results in an error. For example, `set mousescroll="ver:1,ver:2"` is rejected.
* A scroll value of 0 is accepted. Setting a direction to 0 disables scrolling along that axis.
* If a direction is omitted, the default value is used. For example:

`set mousescroll=ver:1,hor:3` - Sets vertical scrolling to 1, horizontal scrolling to 3.
`set mousescroll=ver:1` - Sets vertical scrolling to 1, horizontal scrolling to the default value (6).
`set mousescroll=` - Error, both directions were omitted.

Do we want to go with the string option or the number option from earlier? I'm okay with either, having the ability to control vertical scrolling gets us 95% of the way there. Once it's settled, I'll update the docs and add tests.

**Comment by justinmk** [comment]
All of those choices sound fine. 

> Do we want to go with the string option or the number option from earlier?

not sure what you mean. The `set mousescroll=direction:count,direction:count` form looks great.

File: src/nvim/option.c
Comment (justinmk) [inline src/nvim/option.c:2624]: nit: prefer ternaries for these simple if-else cases, it results in much shorter code without a loss of clarity.

**Review by justinmk — APPROVED** [review]
looks great so far, looking forward to this! 

p.s. For faster feedback our chat channel  https://gitter.im/neovim/neovim can be useful.

File: src/nvim/option.c
Comment (JaySandhu) [reply src/nvim/option.c]: Valid point, I made the change 👍

**Comment by JaySandhu** [comment]
Wrote a test and updated the docs. I think that covers it. The documentation might need some copy editing though. The test works by scrolling and observing changes in the cursor position, figured we could get away with not using `ui.screen`.

File: src/nvim/option.c
Comment (jamessan) [inline src/nvim/option.c:2611]: There's a family of `getdigits` functions in src/nvim/charset.c that encapsulate this sort of functionality and error-checking.  Probably better to use the appropriate one of those.

Comment (JaySandhu) [reply src/nvim/option.c:2611]: I had a look, I think the best way to do it is something like this:
```
string += 4;
*direction = getdigits_int(&string, false, -1);

// Num options are generally kept within the signed int range.
// We know this number won't be negative because we've already checked for
// a minus sign. We'll allow 0 as a means of disabling mouse scrolling.
if (*direction == -1) {
  return e_invarg;
}
```
Idk if that's helping readability, but it's your call.

**Comment by overdrivenpotato** [comment]
Is there something blocking this PR from being merged?

**Comment by dundargoc** [comment]
> Is there something blocking this PR from being merged?

Lack of manpower I'd guess :). @clason: who's most suited to tackle this you think?

**Comment by clason** [comment]
What's blocking is the need for

1. addressing the review comment
2. rebasing on latest `master`

**Comment by JaySandhu** [comment]
Happy to make `getdigits` change and rebase. 

For what it's worth, I don't think using `getdigits` is much of an improvement. Will do what it takes to get this merged though.

**Comment by justinmk** [comment]
@JaySandhu gentle ping

**Comment by JaySandhu** [comment]
Thanks for the ping, this had totally slipped my mind.

- Rebased.
- Squahed commits and reworded the commit message to follow the new guidelines.
- Made the `getdigits` change, in spite of my potestations lol.
- Fixed options.lua, the old commit didn't compile anymore. Changed `defaults={if_true={vi="ver:3,hor:6"}}` to `defaults={if_true="ver:3,hor:6"}`, I think `vi=` isn't a thing anymore? Also, added a `short_desc` (noticed the other options have one now).

File: runtime/doc/options.txt
Comment (zeertzjq) [inline runtime/doc/options.txt:4177]: Double quotes should be used here:
```suggestion
	Direction is one of either "hor" or "ver", "hor" controls horizontal
	scrolling and "ver" controls vertical scrolling. Count sets the amount
```

Comment (JaySandhu) [reply runtime/doc/options.txt]: Made the change. Squashed the commit for cleaner repo history. Thanks for reviewing these old PRs btw 🙂.

**Comment by zeertzjq** [comment]
CI is failing. Looks like this needs a rebase.

**Comment by JaySandhu** [comment]
Rebased and addressed the main CI issue. The type of error messages was changed from `char_u*` to `char*` in https://github.com/neovim/neovim/commit/efa924f66b183d9cf2404ce91c4f009c27e0515a. This PR was still using `char_u*` so the compiler was complaining type mismatches.

I believe the remaining CI failures are unrelated (could be wrong though).

**Comment by JaySandhu** [comment]
Rebased again. Fixed conflicts and linter complaints.

**Review by zeertzjq — APPROVED** [review]

**Comment by xiyaowong** [comment]
> you can no longer scroll inactive windows

Oh, no! Am I the only one who wants to scroll inactive windows?

**Comment by zeertzjq** [comment]
> > you can no longer scroll inactive windows
> 
> Oh, no! Am I the only one who wants to scroll inactive windows?

Then just don't map scroll wheel keys.

**Comment by xiyaowong** [comment]
@zeertzjq I'm sorry, I just tried the latest option and didn't understand it clearly, now I understand

**Pull request merged**
